doc/sphinx/proof-engine/ltac.rst

```diff
@@ -844,66 +961,126 @@ We can carry out pattern matching on terms with:
+           | fun z => ?y * 3 => idtac "y =" y; pose (fun z: nat => y * 5)
            end.
+
+.. _match_vs_lazymatch_ex:
+
+   .. example:: Comparison of match, multimatch and lazymatch
+
+      **todo: how do match and multimatch differ?**
+
+      .. coqtop:: reset in
+
          Goal True.
-         f (3+4).
+
+      .. coqtop:: all
+
+         match False with
+         | False => idtac "try"; fail
+         | _ => easy
+         end.
+
+      .. coqtop:: reset in
+
+         Goal True.
+
+      .. coqtop:: all
+
+         multimatch False with
+         | False => idtac "try"; fail
+         | _ => easy
+         end.
+
+      .. coqtop:: reset in
+
+         Goal True.
+
+      .. coqtop:: all fail
+
+         lazymatch False with
+         | False => idtac "try"; fail
+         | _ => easy
+         end.
 
 .. _ltac-match-goal:
 
-Pattern matching on goals
-~~~~~~~~~~~~~~~~~~~~~~~~~
+Pattern matching on goals: match goal
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-We can perform pattern matching on goals using the following expression:
+.. todo PR The title seems incomplete, should be "goals and hypotheses", right?
 
-.. we should provide the full grammar here
+Use this form to match against goals.  These patterns have zero or
+more subpatterns to match hypotheses followed by a subpattern to match
+the conclusion.
 
-.. tacn:: match goal with {+| {+, @context_hyp} |- @cpattern => @ltac_expr } | _ => @ltac_expr end
-   :name: match goal
+.. tacn:: @match_key {? reverse } goal with {? %| } {+| @match_context_rule } end
+   :name: match goal; lazymatch goal; multimatch goal
 
-   If each hypothesis pattern :n:`hyp`\ :sub:`1,i`, with i = 1, ..., m\ :sub:`1` is
+   .. insertprodn match_context_rule match_hyp
+
+   .. prodn::
+      match_context_rule ::= {*, @match_hyp } |- @match_pattern => @ltac_expr
+      | [ {*, @match_hyp } |- @match_pattern ] => @ltac_expr
+      | _ => @ltac_expr
+      match_hyp ::= @name : @match_pattern
+      | @name := {? [ @match_pattern ] : } @match_pattern
+
+   :tacn:`match goal`, :tacn:`lazymatch goal` and :tacn:`multimatch goal` are :token:`ltac_expr1`\s.
+
+   `multimatch` (in :token:`match_key`)
+      Allows subsequent tactics
+      to backtrack into a right-hand side tactic which has backtracking points
+      left and trigger the selection of a new matching branch or combination of
+      hypotheses when all the backtracking points of the right-hand side have
+      been consumed.
+
+      :n:`match { reverse } goal …` is, in fact, shorthand for
+      :n:`once multimatch { reverse } goal …`.
+
+   `lazymatch` (in :token:`match_key`)
+      Causes the match to commit to the first matching branch with the first
+      matching combination of hypotheses rather than trying a new match if
+      the right-hand side fails. If the right-hand side of the selected branch
+      is a tactic with backtracking points, then subsequent failures cause
+      this tactic to backtrack.
```

Comment:
Since the explanation is basically the same as above (the text is slightly different but the previous explanation was generic enough to apply here as well), I think it could be removed from this section. It would make things much lighter.

Reply:
Perhaps--maybe come back to this after match term is in better shape.  "combination of hypotheses" is definitely different.  Maybe we can link to the descriptions in match terms above.